src/preprocessing.py

```diff
@@ -0,0 +1,443 @@
+    """
+    Detect and extract tables from markdown text and return them along with the cleaned text.
+
+    Args:
+        text (str): The input markdown text containing potential tables.
+
+    Returns:
+        Tuple[List[List[List[str]]], str]:
+            - List[List[List[str]]]: A list of tables, each table is a list of rows, each row is a list of cell strings.
+            - str: The text with tables removed.
+
+    Raises:
+        TypeError: If the input text is not a string.
+    """
+    if not isinstance(text, str):
+        raise TypeError("Input text must be a string.")
+
+    lines = text.split('\n')
+    tables = []
+    non_table_lines = []
+    i = 0
+    while i < len(lines):
+        line = lines[i]
+        # Check if this line contains a pipe '|' and is not a code block line
+        if '|' in line and not line.strip().startswith('```'):
+            # Potential table start
+            table_lines = [line]
+            i += 1
+            if i < len(lines):
+                separator_line = lines[i]
+                # Check if the next line is a table separator line
+                if re.match(r'^\s*\|?\s*(\s*:?-+:?\s*\|)+\s*(:?-+:?\s*)?\s*$', separator_line):
+                    table_lines.append(separator_line)
+                    i += 1
+                    # Collect table body lines
+                    while i < len(lines) and '|' in lines[i] and not lines[i].strip().startswith('```'):
+                        table_lines.append(lines[i])
+                        i += 1
+                    # Add the collected table lines
+                    tables.append(table_lines)
+                    continue  # Skip incrementing i at the end
+                else:
+                    # Not a table separator, add the line to non_table_lines
+                    non_table_lines.append(line)
+                    i -= 1  # Step back to reprocess the separator line
+            else:
+                # End of text, add the line to non_table_lines
+                non_table_lines.append(line)
+        else:
+            # Not a table line, add to non_table_lines
+            non_table_lines.append(line)
+        i += 1
+    # Reconstruct text without tables
+    cleaned_text = '\n'.join(non_table_lines)
+    # Parse tables into structured data
+    parsed_tables = []
+    for table_lines in tables:
+        parsed_table = []
+        for row_line in table_lines:
+            # Skip separator lines
+            if re.match(r'^\s*\|?\s*(\s*:?-+:?\s*\|)+\s*(:?-+:?\s*)?\s*$', row_line):
+                continue
+            # Split the row into cells
+            row = [cell.strip() for cell in row_line.strip().strip('|').split('|')]
+            parsed_table.append(row)
+        parsed_tables.append(parsed_table)
+    return parsed_tables, cleaned_text
+
+def extract_markdown_elements(text: str) -> Tuple[List[str], List[Tuple[str, str]], List[Tuple[str, str]], List[List[List[str]]], str]:
```

Comment:
Thank you for your suggestion regarding the return type of the `extract_markdown_elements` function.

I've decided to keep the current return structure for the following reasons:
1. The method is used only once in our codebase. Adding a new data class for the return types could make things more complex without much benefit because it's not widely used.
2. The return types and values are well documented. This makes it easier to understand the function's output without needing a new structure.

If we start using this function more often in the future, I agree that using a DataFrame or a custom data class might help make our code clearer and easier to manage.